test/test_braid.py

```diff
@@ -329,31 +329,43 @@ def test_braid_relation2_empty(self):
         braid = Braid([])
         with pytest.raises(IllegalTransformationException):
             braid.braid_relation2(index=0)
+    
+    def test_braid_relation2_performable(self):
+        braid = Braid([3, 1, 2, 1])
+        for i in range(-4, 4):
+            assert braid.is_braid_relation2_performable(i) == (i in [-4, -1, 0, 3])
 
     def test_braid_relation2_abs_check1(self):
         braid = Braid([1, 3, -3, 1])
         assert np.all(braid.braid_relation2_performable_indices() == np.array([0, 2]))
 
     def test_braid_relation2_abs_check2(self):
         braid = Braid([1, 3, -1, -2, 4])
-        assert np.all(braid.braid_relation2_performable_indices() == np.array([0, 1, 3]))
+        assert np.all(braid.braid_relation2_performable_indices() == np.array([0, 1, 3, 4]))
 
     def test_braid_relation2_one_element(self):
         braid = Braid([10])
         assert np.all(braid.braid_relation2_performable_indices() == np.array([]))
 
     def test_braid_relation2_negative_exception(self):
         braid = Braid([1, -3, 2])
-        for i in range(1, 5):
+        assert braid.braid_relation2(-3) == Braid([-3, 1, 2])
+        for i in range(-2, 0):
             with pytest.raises(IllegalTransformationException):
-                braid.braid_relation2(-i)
-            
+                braid.braid_relation2(i)
+        for i in range(-5, -3):
```

Comment:
What exactly is the range of the allowable indices? There is only the check for `0 <= index <= len`, so what happens in the other "periods"? (If we want to allow such indices.)

Reply:
See updated docs of braid_relation2; exceptions handled more clearly.